runtime/doc/dev_style.txt

```diff
@@ -859,6 +859,25 @@ execute, simply `assert`: >c
         assert(false);
     }
 
+Switch statements that are conditional on an enumerated value should not have
+a `default` case if it is exhaustive. In cases where an exhaustive switch
+statement on an enum is exhaustive, the compiler might still treat it as if it
+is not exhaustive. This happens in cases where each case contains a return
+statement, but the compiler still complains about there not being a return
+statement for the entire control flow. For situations like these, you should
+use `UNREACHABLE` to explicitly tell the compiler that the default case is
+unreachable. For example: >c
```

Comment:
Hold up. This section makes it sound like you should use `UNREACHABLE` when the compiler gives a warning, rather than all the time regardless of compiler. Is this intended?

Reply:
> Hold up. This section makes it sound like you should use `UNREACHABLE` when the compiler gives a warning, rather than all the time regardless of compiler. Is this intended?

The compiler (GCC and MSVC on CI, that is) will always give a warning in cases where there are return statements for each case but no catch-all return statement

Reply:
That's not my point. The lore on why this rule was created muddies the water on when it should be used, which I assume is "always"?

Reply:
> That's not my point. The lore on why this rule was created muddies the water on when it should be used, which I assume is "always"?

It's "always" in the context of enum switches with a return statement for each case, which is the context of that section of the style guide.

Reply:
I think this style rule is very convoluted and specific. I would seriously reconsider if the benefits it provides are worth it.

Reply:
I do disagree. Control flow checks are very useful, and extra code that masks them into runtime errors are harmful. In general compile time errors should be preferred over runtime errors.

The rule basically says, if the compiler is shit then you need to add this hack. Just because there are some compilers that can't properly check control flow doesn't mean we need to reduce our code to the lowest denominator.